docs/source/extension-module.md

```diff
@@ -94,47 +95,123 @@ if (method_meta.ok()) {
+const auto result = module.execute("forward", tensor);
+```
+
+For the standard `forward()` method, the above can be simplified:
+
+```cpp
+const auto result = module.forward(tensor);
+```
+
+**Note:** `execute()` or `forward()` will load the `Program` and the `Method` the first time they are called. Therefore, the first inference will take longer, as the model is loaded lazily and prepared for execution unless it was explicitly loaded earlier.
+
+### Setting Input and Output
+
+You can set individual input and output values for methods with the following APIs.
 
-### Perform an Inference
+#### Setting Inputs
 
-Assuming that the `Program`'s method names and their input format is known ahead of time, we rarely need to query for those and can run the methods directly by name using the `execute()` function:
+Inputs can be any `EValue`, which includes tensors, scalars, lists, and other supported types. To set a specific input value for a method:
 
 ```cpp
-const auto result = module.execute("forward", Tensor(&tensor));
+module.set_input("forward", input_value, input_index);
 ```
 
-Which can also be simplified for the standard `forward()` method name as:
+- `input_value` is an `EValue` representing the input you want to set.
+- `input_index` is the zero-based index of the input to set.
+
+For example, to set the first input tensor:
 
 ```cpp
-const auto result = module.forward(Tensor(&tensor));
+module.set_input("forward", tensor_value, 0);
 ```
 
-Note: `execute()` or `forward()` will try to force load the `Program` and the `Method` when called for the first time. Therefore, the first inference will take more time than subsequent ones as it loads the model lazily and prepares it for execution unless the `Program` or `Method` was loaded explicitly earlier using the corresponding functions.
+You can also set multiple inputs at once:
+
+```cpp
+std::vector<runtime::EValue> inputs = {input1, input2, input3};
+module.set_inputs("forward", inputs);
+```
+
+**Note:** You can skip the method name argument for the `forward()` method.
+
+By pre-setting all inputs, you can perform an inference without passing any arguments:
+
+```cpp
+const auto result = module.forward();
+```
+
+Or just setting and then passing the inputs partially:
+
+```cpp
+// Set the second input ahead of time.
+module.set_input(input_value_1, 1);
+
+// Execute the method, providing the first input at call time.
+const auto result = module.forward(input_value_0);
+```
+
+**Note:** The pre-set inputs are stored in the `Module` and can be reused multiple times for the next executions.
+
+Don't forget to clear or reset the inputs if you don't need them anymore by setting them to default-constructed `EValue`:
+
+```cpp
+module.set_input(runtime::EValue(), 1);
+```
+
+#### Setting Outputs
+
+Currently, only tensor outputs are supported for setting, and they must not be memory-planned at model export time. Memory-planned tensors are preallocated during model export and cannot be replaced.
```

Comment:
This would be more clear:

"Only outputs of type Tensor can be set at runtime, and ..."

When I first read the current version I interpreted it as "Only outputs of tensors" (which doesn't really mean anything) rather than "only outputs that are tensors".